Add new configuration mechanism for CoreCLR.

adityamandaleeka opened the pull request:
This change allows hosts using `coreclr_initialize` to specify configuration properties via key-value strings. The values of these properties can be retrieved at run time using a new mechanism that allows falling back on CLRConfig settings if desired.

Addresses #3884.

cc: @sergiy-k @jkotas @Maoni0

jkotas commented on src/vm/eeconfig.cpp:
```diff
@@ -789,10 +792,18 @@ HRESULT EEConfig::sync()
-    int forceGCconcurrent = CLRConfig::GetConfigValue(CLRConfig::UNSUPPORTED_gcConcurrent);
-    if ((forceGCconcurrent > 0) || (forceGCconcurrent == -1 && g_IGCconcurrent))
+    bool gcConcurrentWasForced = false;
+#ifdef FEATURE_CORECLR
```
I do not think we need this change and the related change in clrconfig.h.

What will not  work if both these changes are reverted?

adityamandaleeka replied:
If we revert the change in clrconfigvalues.h, the startup flag for concurrent GC will not be enabled unless "System.GC.Concurrent" is explicitly set in the JSON configuration. We can change the code (in unixinterface.cpp) that sets the flag so that it uses the boolean argument fallback instead of the COMPlus one and always returns true if the value isn't explicitly set, but then you won't be able to use the COMPlus mechanism to configure the concurrent GC startup flag.

If we keep the change in clrconfigvalues.h, the existing code that checks for -1 doesn't make sense anymore for CoreCLR.

jkotas replied:
We are looking up the `CLRConfig::UNSUPPORTED_gcConcurrent` and `System.GC.Concurrent` both here as well as in `InitializeStartupFlags`. It is redundant.

I think we should either:

- Lookup both in `InitializeStartupFlags` and do nothing here (put this piece of code here under `#ifndef FEATURE_CORECLR` ifdef)

or

- Lookup just `System.GC.Concurrent` in `InitializeStartupFlags` (use the overload that takes default value as constant) and `UNSUPPORTED_gcConcurrent` here - the existing code here should work just fine for it, no need to ifdef it.

adityamandaleeka replied:
We can't simply ifdef it out entirely because a bit further down, we need to know whether concurrent GC was forced. If it wasn't, we disable concurrent GC if GCStress is enabled.

However, doing the second option does mean that we won't treat the presence of `"System.GC.Concurrent": true` as the user forcing concurrent GC (in other words, we may turn it off even if they specify it).

If we're okay with that, I can undo the change in clrconfigvalues.h and here, and change `InitializeStartupFlags` to fallback to 'true' if "System.GC.Concurrent" isn't set.

sergiy-k replied:
I would prefer to keep the existing behavior. I think that we do want to be able run GCStress with Concurrent GC enabled.

Maoni0 replied:
We definitely should allow the option to run GCStress with concurrent GC enabled - at the very least, you should be able to turn on/off concurrent GC for GCStress 3. 

what is the order of evaluation supposed to be here? complus overwrites startup flag which overwrites config?

adityamandaleeka replied:
> what is the order of evaluation supposed to be here? complus overwrites startup flag which overwrites config?

With my changes, we will first determine whether the startup flag for concurrent GC should be set with this criteria:

- If the COMPlus value is set explicitly, we will respect that (treating 1 as true and anything else as false).
- If not, we will check whether the user has set "System.GC.Concurrent" explicitly to true or false in the config file (in the case of CLI this is the JSON configuration) and respect that.
- If neither of them is set explicitly, we will use the default specified in the ConfigDWORDInfo, which for CoreCLR will be 1 (true). 

If we go through all that and decide concurrent GC should be on, we will treat that as 'force' enabling. So with no special configuration set for concurrent GC and GCStress on, we will have both enabled. You can turn off concurrent GC by either setting "System.GC.Concurrent" to false in your JSON config or setting the COMPlus value to 0 (or anything other than 1).

jkotas replied:
> treating 1 as true and anything else as false

Pretty much all existing COMPlus settings treat 0 as false, and everything else as true

adityamandaleeka replied:
Ah ok, I will change that in configuration.cpp in my next commit.

jkotas commented on src/utilcode/configuration.cpp:
```diff
@@ -0,0 +1,148 @@
+    LPCWSTR knobValue = GetConfigurationValue(name);
+    if (knobValue != nullptr)
+    {
+        return (wcscmp(knobValue, W("true")) == 0);
```
Would it make sense to also allow 0/1 as alternative to true/false?

adityamandaleeka replied:
I think if we're already expecting an exact string for the property key, it's not any more painful to specify the exact value as well. 

As an added bonus, for hosts that use JSON for these configuration parameters, 'true' and 'false' are the idiomatic way of expressing boolean values.

sergiy-k replied:
I think we should support the values that make sense for the runtime confing JSON file. We can always extend it if needed.

jkotas commented on src/utilcode/configuration.cpp:
```diff
@@ -0,0 +1,148 @@
+        if (names[i] != nullptr && values[i] != nullptr)
+        {
+            size_t bufferLength = wcslen(names[i]) + 1;
+            LPWSTR nameBuffer = new WCHAR[bufferLength];
```
The caller just created a copy. You should not need to create another one here - just take ownership of the copy created by the caller.

jkotas commented on src/utilcode/clrconfig.cpp:
```diff
@@ -317,7 +326,11 @@ DWORD CLRConfig::GetConfigValue(const ConfigDWORDInfo & info)
 
-    DWORD result = info.defaultValue;
+    _ASSERT(result != nullptr);
+    if (result == nullptr)
```
This should not be necessary.

jkotas commented on src/utilcode/clrconfig.cpp:
```diff
@@ -297,16 +297,25 @@ BOOL CLRConfig::IsConfigEnabled(const ConfigDWORDInfo & info)
+// 
 // static
-DWORD CLRConfig::GetConfigValue(const ConfigDWORDInfo & info)
+bool CLRConfig::GetConfigValue(const ConfigDWORDInfo & info, bool useDefaultIfNotSet, bool acceptExplicitDefaultFromRegutil, DWORD *result)
```
Instead of making this method even more complex than what it is, it would be better to create artificial `CLRConfig::ConfigDWORDInfo` and use this function without any changes - similar how `JitHost::getIntConfigValue` does it.

adityamandaleeka replied:
I don't think this method has really been made more complex. In fact, there were some comments in there about how it previously wasn't treating explicitly defined default values coming from REGUTIL properly. Those areas have now been cleaned up, and the switch was added to preserve existing behavior for backwards compatibility.

If we were to change it back, we'd have to always check whether the (JSON) property key was set, and then pass in a ConfigDWORDInfo which will have a different value for the default based on whether the JSON value was set or not. And in the end, we'll be back in the situation where callers to GetConfigValue can't know whether the returned value was actually set or whether it was a default.

I think this change is worth keeping. However, if you feel strongly about it, I'm willing to change it.

jkotas replied:
I do not think this change is good - it feels way more complex than what it needs to be.

You just need to know whether the value was actually set or whether it is the default. A single `[out] bool *` argument should be sufficient to give you that.

adityamandaleeka commented:
I've updated the PR based on feedback and also updated the code reading ThreadPool_Force[Min|Max]WorkerThreads to use the new system.

Please let me know if there are any further comments.

jkotas commented on src/vm/win32threadpool.cpp:
```diff
@@ -392,11 +393,11 @@ BOOL ThreadpoolMgr::Initialize()
     // initialize Worker and CP thread settings
     DWORD forceMin;
-    forceMin = CLRConfig::GetConfigValue(CLRConfig::INTERNAL_ThreadPool_ForceMinWorkerThreads);
+    forceMin = Configuration::GetKnobDWORDValue(W("System.Threading.ThreadPool.MinThreads"), CLRConfig::INTERNAL_ThreadPool_ForceMinWorkerThreads);
```
Could you please factor this out into a small helper method, so that the actual Configuration call is done in just single place?

adityamandaleeka replied:
Sure, done. I added helper functions for getting min/max. For now, the behavior will be the same though (each check results in a lookup). I think we should add a caching mechanism like we have for the old CLRConfig types, but that can be a separate change (not rushed for RC2).

jkotas commented on src/utilcode/configuration.cpp:
```diff
@@ -0,0 +1,145 @@
+
+void Configuration::InitializeConfigurationKnobs(int numberOfConfigs, LPCWSTR *names, LPCWSTR *values)
+{
+    knobs = new (nothrow) ConfigurationKnob[numberOfConfigs];
```
This copy should not be necessary.

jkotas replied:
(Just remember the names/values arrays)

jkotas commented on src/utilcode/clrconfig.cpp:
```diff
@@ -297,16 +297,25 @@ BOOL CLRConfig::IsConfigEnabled(const ConfigDWORDInfo & info)
 // static
-DWORD CLRConfig::GetConfigValue(const ConfigDWORDInfo & info)
+DWORD CLRConfig::GetConfigValue(const ConfigDWORDInfo & info, bool acceptExplicitDefaultFromRegutil, /* [Out] */ bool *isDefault)
 {
```
I do not understand why we need `acceptExplicitDefaultFromRegutil`. What will not work without it?

adityamandaleeka replied:
Effectively, there was a bug in the old code where it would reject a value that came from REGUTIL if it matched the default value, even if it was explicitly set. The comment near that code explained that they left it that way for legacy compat reasons (in case someone was depending on that behavior). Since there are too many users of this function to investigate on a case-by-case basis, this switch allows us to maintain the legacy behavior for all the existing callers, and for new areas that we're touching we can use the fixed behavior.

It also means that the new configuration stuff really does what it says (uses an explicit value if it's set, even if it matches a default). You can imagine a scenario where "System.Foo.Whatever" is set to something other than the default. If someone wants to override that and set it back to the default with a REGUTIL setting, they couldn't do that without 'acceptExplicitDefaultFromRegutil'.

adityamandaleeka commented on src/dlls/mscoree/unixinterface.cpp:
```diff
@@ -86,76 +87,47 @@ static LPCWSTR* StringArrayToUnicode(int argc, LPCSTR* argv)
+            STARTUP_FLAGS::STARTUP_LOADER_OPTIMIZATION_SINGLE_DOMAIN |
+            STARTUP_FLAGS::STARTUP_SINGLE_APPDOMAIN);
+
+    if (Configuration::GetKnobBooleanValue(W("System.GC.Concurrent"), CLRConfig::UNSUPPORTED_gcConcurrent))
```
I think we don't really need to set this flag anymore. For CoreCLR, we're going to default to concurrent GC enabled anyway, so the flag doesn't really do anything. To disable concurrent GC, the user will either set "System.GC.Concurrent" to false or set the COMPlus value to false.

Any thoughts?

jkotas commented on src/utilcode/clrconfig.cpp:
```diff
@@ -362,46 +372,64 @@ DWORD CLRConfig::GetConfigValue(const ConfigDWORDInfo & info)
+            // If we are willing to accept the default value when it's set explicitly,
+            // checking the HRESULT here is sufficient. E_FAIL is returned when the
+            // default is used.
+            if (hr != E_FAIL)
```
This should better be `if (!FAILED(hr))`

jkotas commented on src/utilcode/clrconfig.cpp:
```diff
@@ -409,14 +437,30 @@ DWORD CLRConfig::GetConfigValue(const ConfigDWORDInfo & info)
+            // checking the HRESULT here is sufficient. E_FAIL is returned when the
+            // default is used.
+            if (hr != E_FAIL)
+            {
```
Same here

jkotas commented on src/utilcode/configuration.cpp:
```diff
@@ -0,0 +1,124 @@
+static LPCWSTR GetConfigurationValue(LPCWSTR name)
+{
+    _ASSERT(name != nullptr);
+    if (name == nullptr || knobNames == nullptr || knobValues == nullptr)
```
These null checks should not be needed

jkotas commented:
LGTM modulo comments

adityamandaleeka commented:
Thanks @jkotas. Comments have been addressed.

sergiy-k commented:
@dotnet-bot test Windows_NT x86 Release Legacy Backend Build and Test  please

sergiy-k commented:
LGTM
